mesosdriver/executor_driver.go

```diff
@@ -0,0 +1,318 @@
+func (driver *ExecutorDriver) sendFailedTasks() {
+	for taskID, status := range driver.failedTasks {
+		updateErr := driver.SendStatusUpdate(status)
+		if updateErr != nil {
+			log.Warnf(
+				"failed to send status update for task %s: %+v", taskID.Value, updateErr,
+			)
+		} else {
+			delete(driver.failedTasks, taskID)
+		}
+	}
+}
+
+// SendStatusUpdate takes a new Mesos status and relays it to the agent
+func (driver *ExecutorDriver) SendStatusUpdate(status mesos.TaskStatus) error {
+	upd := calls.Update(status)
+	resp, err := driver.cli.Send(context.TODO(), calls.NonStreaming(upd))
+	if resp != nil {
+		resp.Close()
+	}
+	if err != nil {
+		log.Errorf("failed to send update: %+v", err)
+		logDebugJSON(upd)
+	} else {
+		driver.unackedUpdates[string(status.UUID)] = *upd.Update
+	}
+	return err
+}
+
+// marshalJSON is a narrowly scoped interface used to allow logDebugJSON to
+// properly format most Mesos messages.
+type marshalJSON interface {
+	MarshalJSON() ([]byte, error)
+}
+
+// logDebugJson prints failed messages to the logger when we can't talk to the
+// Agent correctly.
+func logDebugJSON(mk marshalJSON) {
+	b, err := mk.MarshalJSON()
+	if err == nil {
+		log.Debug(string(b))
+	}
+}
+
+// NewStatus returns a properly configured Mesos.TaskStatus
+func (driver *ExecutorDriver) NewStatus(id mesos.TaskID) mesos.TaskStatus {
+	return mesos.TaskStatus{
+		TaskID:     id,
+		Source:     mesos.SOURCE_EXECUTOR.Enum(),
+		ExecutorID: &driver.executor.ExecutorID,
+		UUID:       []byte(uuid.NewRandom()),
+	}
+}
+
+func (driver *ExecutorDriver) Stop() {
+	driver.shouldQuit = true
+}
+
+// NewExecutorDriver returns a properly configured ExecutorDriver
+func NewExecutorDriver(mesosConfig *config.Config, delegate TaskDelegate) *ExecutorDriver {
+	agentApiUrl := url.URL{
+		Scheme: "http",
+		Host:   mesosConfig.AgentEndpoint,
+		Path:   agentAPIPath,
+	}
+
+	http := httpcli.New(
+		httpcli.Endpoint(agentApiUrl.String()),
+		httpcli.Codec(codecs.ByMediaType[codecs.MediaTypeProtobuf]),
+		httpcli.Do(httpcli.With(httpcli.Timeout(driverHttpTimeout))),
+	)
+
+	callOptions := executor.CallOptions{
+		calls.Framework(mesosConfig.FrameworkID),
+		calls.Executor(mesosConfig.ExecutorID),
+	}
+
+	driver := &ExecutorDriver{
+		cli: calls.SenderWith(
+			httpexec.NewSender(http.Send),
+			callOptions...,
+		),
+		unackedTasks:   make(map[mesos.TaskID]mesos.TaskInfo),
+		unackedUpdates: make(map[string]executor.Call_Update),
+		failedTasks:    make(map[mesos.TaskID]mesos.TaskStatus),
+		delegate:       delegate,
+		cfg:            mesosConfig,
+	}
+
+	driver.subscriber = calls.SenderWith(
+		httpexec.NewSender(http.Send, httpcli.Close(true)),
+		callOptions...,
+	)
+
+	return driver
+}
+
+// Run makes sure we're subscribed to events, and restarts the event loop until
+// we're told to stop.
+func (driver *ExecutorDriver) Run() error {
+	shouldReconnect := driver.maybeReconnect()
+
+	disconnectTime := time.Now()
+	handler := driver.buildEventHandler()
+
+	for {
```

Comment:
The structure of this is kinda funky, but it's essentially a reconnect loop that will make sure the driver stays connected to the Agent. This is required now because the Executor is expected to maintain a constant HTTP connection to the agent. It also must track whethere all its requests were acknowledged or not since it might have been disconnected. Thus there is some state management throughout here.